### PR TITLE
service/apt: Implement soft reset & CloseApplication

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Disclaimer: I'm not sure if this is the correct way to do this, and  I don't know what is the correct way to do it either. So this may be bad impl. If that's the case, I'll just close.

With this, emulation can properly reset when exiting from Mystery Gift menu in Pokemon US/UM/OR/AS. X/Y doesn't work, as it seems to hang before calling service functions (I'm not sure). It can also properly stop & return to game list when exiting from Checkpoint.

As @jroweboy stated, I'm not sure what a Reset should do. Currently I think shutting down the system and then restarting is good enough, but the 3ds may have a different way of doing this. But I do think this is similar to expected behavior, and should not cause any problems, as it is actually a clean reload of the system.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3959)

### CODE REVIEW

File: src/core/hle/service/apt/apt.cpp [hunk 1]
```diff
@@ -542,6 +542,66 @@ void Module::Interface::StartLibraryApplet(Kernel::HLERequestContext& ctx) {
...
+        // to soft reset.
+        application_reset_prepared = true;
+    }
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+    rb.Push(RESULT_SUCCESS);
+}
+
+void Module::Interface::DoApplicationJump(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x32, 4, 2);
```
Comment (wwylele) [inline src/core/hle/service/apt/apt.cpp]: The ID `0x00320084` should be translated as `0x32, 2, 4` here (

File: src/core/hle/service/apt/apt.cpp [hunk 2]
```diff
@@ -542,6 +542,66 @@ void Module::Interface::StartLibraryApplet(Kernel::HLERequestContext& ctx) {
...
+    u32 buffer1_size = rp.Pop<u32>();
+    u32 buffer2_size = rp.Pop<u32>();
+    u32 flag = rp.Pop<u32>();
```
Comment (wwylele) [inline src/core/hle/service/apt/apt.cpp]: Where does this come from? It is mentioned in 3dbrew or the comment and it is conflict with parameter count in the ID

File: src/core/hle/service/apt/apt.cpp [hunk 3]
```diff
@@ -542,6 +542,66 @@ void Module::Interface::StartLibraryApplet(Kernel::HLERequestContext& ctx) {
...
+    std::vector<u8> buffer1 = rp.PopStaticBuffer();
+    std::vector<u8> buffer2 = rp.PopStaticBuffer();
```
Comment (wwylele) [inline src/core/hle/service/apt/apt.cpp]: Name these buffers according to the doc comment.

File: src/core/hle/service/apt/apt.cpp [hunk 4]
```diff
@@ -542,6 +542,66 @@ void Module::Interface::StartLibraryApplet(Kernel::HLERequestContext& ctx) {
...
+
+    Core::System::GetInstance().RequestShutdown();
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+    rb.Push(RESULT_SUCCESS);
+}
+
+void Module::Interface::PrepareToDoApplicationJump(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x31, 4, 0);
+    u32 flags = rp.Pop<u32>();
```
Comment (wwylele) [inline src/core/hle/service/apt/apt.cpp]: u8 according to 3dbrew.

File: src/core/hle/service/apt/apt.cpp (hunk 1 above)
Comment (zhaowenlan1779) [reply src/core/hle/service/apt/apt.cpp]: Oh, correct

File: src/core/hle/service/apt/apt.cpp (hunk 2 above)
Comment (zhaowenlan1779) [reply src/core/hle/service/apt/apt.cpp]: sorry, I think I read the wrong page

**Comment by zhaowenlan1779** [comment]
@wwylele addressed

**Comment by wwylele** [comment]
A proper implementation should create a new KProcess object and transfer application status/rights(?) to it. This proper implementation is also needed when we implement launching game from home menu. However the proper implementation touches many modules (multiple services apt/ns/am/pm, ncch loader/file system, and probably rasterizer cache problem when switching process = also about memory model rewrite...), I can let this idea pass for now as a workaround method before all these parts are tackled.

**Comment by Leo626** [comment]
Tested Pokemon X/Y. Still hangs at the same lines (see #3703). 

```
[  12.027384] HW.Memory <Error> core\memory.cpp:Memory::GetPhysicalPointer:292: MMIO mappings are not supported yet. phys_addr=0x10B00000
[  12.027389] HW.GPU <Critical> core\hw\gpu.cpp:GPU::MemoryFill:87: invalid end address 0X10B00000
[  12.043915] Service <Error> core\hle\service\service.cpp:Service::ServiceFrameworkBase::ReportUnimplementedFunction:161: unknown / unimplemented function 'ReplySleepQuery': port='APT:U' cmd_buf={[0]=0x3e0080, [1]=0x300, [2]=0x0}
```

**Comment by zhaowenlan1779** [comment]
@Leo626 
I suggest they use a different function than later games.
Too bad..

**Comment by jroweboy** [comment]
I think this is an interesting stop gap solution to the soft reset issue, but just as a personal opinion, i don't think its worth the additional hacked up communication between frontend and core. I would be very accepting of a full solution that completes what @wwylele mentioned, or if the core and frontend communication wasn't as convulted as it currently is, such that adding a "reset pls" state was much simpler.

Thanks for the contribution!

**Comment by zhaowenlan1779** [comment]
from Discord
```
jroweboy Yesterday at 9:08 AM
so i guess i really don't mind the soft reset workaround @zhaowenlan1779
```

**Comment by B3n30** [comment]
resolve conflicts, please

**Pull request merged**